test/test_http2_connection.py

```diff
@@ -0,0 +1,327 @@
+        assert not _is_illegal_header_value(b"foo"), "foo"
+        assert not _is_illegal_header_value(b"foo bar"), "foo bar"
+        assert not _is_illegal_header_value(b"foo\tbar"), "foo\\tbar"
+
+        # A field value MUST NOT contain the zero value (ASCII NUL, 0x00), line
+        # feed (ASCII LF, 0x0a), or carriage return (ASCII CR, 0x0d) at any
+        # position. [1]
+        assert _is_illegal_header_value(b"foo\0bar"), "foo\\0bar"
+        assert _is_illegal_header_value(b"foo\x00bar"), "foo\\x00bar"
+        assert _is_illegal_header_value(b"foo\x0abar"), "foo\\x0abar"
+        assert _is_illegal_header_value(b"foo\x0dbar"), "foo\\x0dbar"
+        assert _is_illegal_header_value(b"foo\rbar"), "foo\\rbar"
+        assert _is_illegal_header_value(b"foo\nbar"), "foo\\nbar"
+        assert _is_illegal_header_value(b"foo\r\nbar"), "foo\\r\\nbar"
+
+        # A field value MUST NOT start or end with an ASCII whitespace character
+        # (ASCII SP or HTAB, 0x20 or 0x09). [1]
+        assert _is_illegal_header_value(b" foo"), " foo"
+        assert _is_illegal_header_value(b"foo "), "foo "
+        assert _is_illegal_header_value(b"foo\x20"), "foo\\x20"
+        assert _is_illegal_header_value(b"\tfoo"), "\\tfoo"
+        assert _is_illegal_header_value(b"foo\t"), "foo\\t"
+        assert _is_illegal_header_value(b"foo\x09"), "foo\\x09"
+
+    def test_default_socket_options(self) -> None:
+        conn = HTTP2Connection("example.com")
+        assert conn.socket_options == [(socket.IPPROTO_TCP, socket.TCP_NODELAY, 1)]
+        assert conn.port == 443
+
+    def test_putheader(self) -> None:
+        conn = HTTP2Connection("example.com")
+        conn.putheader("foo", "bar")
+        assert conn._headers == [(b"foo", b"bar")]
+
+    def test_request_putheader(self) -> None:
+        conn = HTTP2Connection("example.com")
+        conn.sock = mock.MagicMock(
+            sendall=mock.Mock(return_value=None),
+        )
+        conn.putheader = mock.MagicMock(return_value=None)  # type: ignore[method-assign]
+        conn.request("GET", "/", headers={"foo": "bar"})
+        conn.putheader.assert_has_calls(
+            [
+                mock.call("foo", "bar"),
+                mock.call(b"user-agent", _get_default_user_agent()),
+            ]
+        )
+
+    def test_putheader_ValueError(self) -> None:
+        conn = HTTP2Connection("example.com")
+        with pytest.raises(ValueError):
+            conn.putheader("foo\0bar", "baz")
+        with pytest.raises(ValueError):
+            conn.putheader("foo", "foo\r\nbar")
+
+    def test_endheaders_ConnectionError(self) -> None:
+        conn = HTTP2Connection("example.com")
+        with pytest.raises(ConnectionError):
+            conn.endheaders()
+
+    def test_send_ConnectionError(self) -> None:
+        conn = HTTP2Connection("example.com")
+        with pytest.raises(ConnectionError):
+            conn.send(b"foo")
+
+    def test_send_bytes(self) -> None:
```

Comment:
For these tests which mock `_h2_conn` internals, I think we should avoid doing this sort of testing since we don't "own" the internals of H2Connection. Instead I think we can get full coverage by making live requests (using the `http_version` fixture mentioned above) to be able to get test coverage without mocks.

You can keep all the tests that don't mock the h2 internals, those look great! :rocket:

Reply:
AFAIK I'm not actually going down below the `h2` public API, just verifying that the method signatures we create are correct. `conn._h2_conn._obj` is our instance of `h2.connection.H2Connection`. These did get elaborate in the pursuit of 100% coverage. 

I get what you're saying about the `http_version` fixture now and use that in my latest commit.

Reply:
Oh gotcha! I forgot we maintained our own structure for that, thank you!